More updates. (#168)
brevzin/cpp_proposals: 2996_reflection/reflection.md

  - any value with _structural type_ that is a permitted result of a constant expression
  - the null reflection (when default-constructed)

We for now restrict the space of reflectable values to those of structural type in order to meet two requirements:

1. The compiler must know how to mangle any reflectable value (i.e., when a reflection thereof is used as a template argument).
2. The compiler must know how to compare any two reflectable values, ideally without interpreting user-defined comparison operators (i.e., to implement comparison between reflections).

> Values of structural types can already be used as template arguments (so implementations must already know how to mangle them), and the notion of _template-argument-equivalent_ values defined on the class of structural types helps guarantee that `&fn<^value1> == &fn<^value2>` if and only if `&fn<value1> == &fn<value2>`.

Reviewer: Perfect, thanks.

### Freestanding implementations

Several important metafunctions, such as `std::meta::nonstatic_data_members_of`, return a `std::vector` value.

> Unfortunately, that means that they are currently not usable in a freestanding environment, but a recent paper ([@P3295R0]) currently being seen by SG7 proposes freestanding `std::vector`, `std::string`, and `std::allocator` in constant evaluated contexts, explicitly to make the facilities proposed by this paper work in freestanding.

Reviewer: I think it's enough to just point out the paper - the phrase "currently being seen by SG7" isn't true yet, and will only have been true for a very short period of time (if at all... probably it should go straight to LEWG?) 

```suggestion
Unfortunately, that means that they are currently not usable in a freestanding environment, but [@P3295R0] currently proposes freestanding `std::vector`, `std::string`, and `std::allocator` in constant evaluated contexts, explicitly to make the facilities proposed by this paper work in freestanding.
```

* [*.#]{.pnum} Otherwise, if both operands are reflections of a namespace alias, alias template, or type alias, then they compare equal if their reflected aliases share the same name, are declared within the same enclosing scope, and alias the same underlying entity.
* [*.#]{.pnum} Otherwise, if neither operand is a reflection of a value, then they compare equal if they are reflections of the same entity.
* [*.#]{.pnum} Otherwise, if one operand is a reflection of a value and the other is not, then they compare unequal.

> * [*.#]{.pnum} Otherwise, if both operands are reflections of values, then they compare equally if and only if the reflected values are _template-argument-equivalent_ ([temp.type]{.sref}).

Reviewer: 👍

```cpp
consteval bool is_incomplete_type(info r);
```
[#]{.pnum} *Mandates*: `r` is a reflection designating a type.

[#]{.pnum} *Returns*: `false` if the type designated by `dealias(r)` is a complete class type. Otherwise, `true`.

> [#]{.pnum} *Effects*: If `dealias(r)` designates a class template specialization with a reachable definition, the specialization has been instantiated.

Reviewer: Wrong tense:
```suggestion
[#]{.pnum} *Effects*: If `dealias(r)` designates a class template specialization with a reachable definition, the specialization is instantiated.
```

  consteval vector<info> members_of(info r, Fs... filters);
```

[#]{.pnum} *Mandates*: `r` is a reflection designating either a complete class type or a namespace and `(std::predicate<Fs, info> && ...)` is `true`.

[#]{.pnum} *Returns*: A `vector` containing the reflections of all the direct members `m` of the entity, excluding any structured bindings, designated by `r` such that `(filters(m) && ...)` is `true`.
Non-static data members are indexed in the order in which they are declared, but the order of other kinds of members is unspecified. [Base classes are not members.]{.note}

> [#]{.pnum} *Effects*: If `dealias(type)` designates a class template specialization with a reachable definition, the specialization has been instantiated.

Reviewer: Ditto:
```suggestion
[#]{.pnum} *Effects*: If `dealias(type)` designates a class template specialization with a reachable definition, the specialization is instantiated.
```

  consteval vector<info> bases_of(info type, Fs... filters);
```

[#]{.pnum} *Mandates*: `type` is a reflection designating a complete class type and `(std::predicate<Fs, info> && ...)` is `true`.

[#]{.pnum} *Returns*: Let `C` be the type designated by `type`. A `vector` containing the reflections of all the direct base classes `b`, if any, of `C` such that `(filters(b) && ...)` is `true`.
The base classes are indexed in the order in which they appear in the *base-specifier-list* of `C`.

> [#]{.pnum} *Effects*: If `dealias(type)` designates a class template specialization with a reachable definition, the specialization has been instantiated.

Reviewer: Ditto:
```suggestion
[#]{.pnum} *Effects*: If `dealias(type)` designates a class template specialization with a reachable definition, the specialization is instantiated.
```

[#]{.pnum} *Effects*: Equivalent to: `return members_of(type, is_variable);`

```cpp
consteval vector<info> accessible_static_data_members_of(info type);
```

[#]{.pnum} *Mandates*: `type` is a reflection designating a complete class type.

[#]{.pnum} *Effects*: Equivalent to: `return members_of(type, is_variable, is_accessible);`

```cpp
consteval vector<info> nonstatic_data_members_of(info type);
```

[#]{.pnum} *Mandates*: `type` is a reflection designating a complete class type.

[#]{.pnum} *Effects*: Equivalent to: `return members_of(type, is_nonstatic_data_member);`

```cpp
consteval vector<info> accessible_nonstatic_data_members_of(info type);
```

[#]{.pnum} *Mandates*: `type` is a reflection designating a complete class type.

[#]{.pnum} *Effects*: Equivalent to: `return members_of(type, is_nonstatic_data_member, is_accessible);`

```cpp
consteval vector<info> subobjects_of(info type);
```

[#]{.pnum} *Mandates*: `type` is a reflection designating a complete class type.

[#]{.pnum} *Returns*: A `vector` containing all the reflections in `bases_of(type)` followed by all the reflections in `nonstatic_data_members_of(type)`.

> [#]{.pnum} *Effects*: If `dealias(type)` designates a class template specialization with a reachable definition, the specialization has been instantiated.

Reviewer: Ditto:
```suggestion
[#]{.pnum} *Effects*: If `dealias(type)` designates a class template specialization with a reachable definition, the specialization is instantiated.
```

```cpp
consteval vector<info> accessible_subobjects_of(info type);
```

[#]{.pnum} *Mandates*: `type` is a reflection designating a complete class type.

[#]{.pnum} *Returns*: A `vector` containing all the reflections in `accessible_bases_of(type)` followed by all the reflections in `accessible_nonstatic_data_members_of(type)`.

> [#]{.pnum} *Effects*: If `dealias(type)` designates a class template specialization with a reachable definition, the specialization has been instantiated.

Reviewer: Ditto:
```suggestion
[#]{.pnum} *Effects*: If `dealias(type)` designates a class template specialization with a reachable definition, the specialization is instantiated.
```
